pkg/operator/controller/externaldns/deployment.go

```diff
@@ -190,14 +193,38 @@ func desiredExternalDNSDeployment(namespace, image, secretName string, serviceAc
 		isOpenShift:    isOpenShift,
 		platformStatus: platformStatus,
 	}
+
 	if len(externalDNS.Spec.Zones) == 0 {
-		container, err := cbld.build("")
-		if err != nil {
-			return nil, fmt.Errorf("failed to build container: %w", err)
+		// an empty list means publish to all zones
+		// this is a special case for Azure
+		// both public and private zones will need to be published to
+		if cbld.provider == externalDNSProviderTypeAzure {
+			providerList := []string{externalDNSProviderTypeAzure, externalDNSProviderTypeAzurePrivate}
+			for _, p := range providerList {
+				cbld.provider = p
+				container, err := cbld.build("")
+				if err != nil {
+					return nil, fmt.Errorf("failed to build container: %w", err)
+				}
+				depl.Spec.Template.Spec.Containers = append(depl.Spec.Template.Spec.Containers, *container)
+			}
+		} else {
+			container, err := cbld.build("")
+			if err != nil {
+				return nil, fmt.Errorf("failed to build container: %w", err)
+			}
+			depl.Spec.Template.Spec.Containers = []corev1.Container{*container}
 		}
-		depl.Spec.Template.Spec.Containers = []corev1.Container{*container}
 	} else {
 		for _, zone := range externalDNS.Spec.Zones {
+			// check each zone for the keyword 'privatednszones', if the provider is Azure
+			// this ensures that the provider 'azure-private-dns' is passed to the container
+			// to set the operand provider correctly
+			if cbld.provider == externalDNSProviderTypeAzure {
+				if strings.Contains(strings.ToLower(zone), azurePrivateDNSZonesResourceSubStr) {
+					cbld.provider = externalDNSProviderTypeAzurePrivate
+				}
+			}
```

Comment:
This can be checked inside the container builder as the externalDNS instance is passed to it.
This way we would keep all the platform specific logic in the container builder.

Reply:
Added the check in cbld.build()